File: 24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        lenh = 0
        l1 = head
        if head == None:
            return None
        elif head.next ==  None:
            return head
        else:
            D = head
            G = head
            S = head.next
            T = head.next.next
            G.next = T
            S.next = G
            D = S
            head = D
            while(G != None) and (S != None) and (T != None):
                if S.next.next.next == None:
                    break
                else:
                    T = T.next.next
                    S = S.next.next.next
                    G = G.next
                    D = D.next
                    D.next = None
                    G.next = T
                    S.next = G
                    D.next = S
                    D = D.next
            return head
                    
        
            
            

File: 24. Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head -> next == NULL)
            return head;
        ListNode *p1 = head, *p2 = head, *p3 = head -> next, *p4 = head -> next -> next;
        p2 -> next = p4;
        p3 -> next = p2; 
        p1 = p3;
        head = p1; //change first two nodes
        while (p2 != NULL && p3 != NULL && p4 != NULL){
        if (p3 -> next -> next -> next != NULL){
            p1 = p1 -> next;
            p2 = p2 -> next;
            p3 = p3 -> next -> next -> next;
            p4 = p4 -> next -> next;
            p1 -> next = NULL;
            p2 -> next = p4;
            p3 -> next = p2;
            p1 -> next = p3;
            p1 = p1 -> next;
        } 
        else
            break;
        }
        return head;
    }
};
